packages/dma-contracts/contracts/core/OperationExecutor.sol

```diff
@@ -126,83 +137,111 @@ contract OperationExecutor is IERC3156FlashBorrower, IFlashLoanRecipient {
     uint256 fee,
     bytes calldata data
   ) external override returns (bytes32) {
+    checkIfFlashloanIsInProgress();
+
     FlashloanData memory flData = abi.decode(data, (FlashloanData));
-    address lender = registry.getRegisteredService(FLASH_MINT_MODULE);
 
-    checkIfLenderIsTrusted(lender);
+    address mcdFlash = CHAINLOG_VIEWER.getServiceAddress(MCD_FLASH);
+    checkIfLenderIsTrusted(mcdFlash);
     checkIfFlashloanedAssetIsTheRequiredOne(asset, flData.asset);
     checkIfFlashloanedAmountIsTheRequiredOne(asset, flData.amount);
 
     processFlashloan(flData, initiator);
 
-    uint256 paybackAmount = amount.add(fee);
-    require(
-      IERC20(asset).balanceOf(address(this)) >= paybackAmount,
-      "Insufficient funds for payback"
-    );
+    uint256 paybackAmount = amount + fee;
+    uint256 funds = IERC20(asset).balanceOf(address(this));
+    if (funds < paybackAmount) {
+      revert InsufficientFunds(funds, paybackAmount);
+    }
+
+    IERC20(asset).safeApprove(mcdFlash, paybackAmount);
 
-    IERC20(asset).safeApprove(lender, paybackAmount);
     return keccak256("ERC3156FlashBorrower.onFlashLoan");
```

Comment:
We could cache the value of this `keccak256` in the constructor as immutable and then return it here, saving some gas